### PR TITLE
feat(watch): reload rules on ConfigMap volume updates

### PR DESCRIPTION
**Pull request by gitlayzer**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it:**

The rules watcher only registered a watch on the rules **file inode** and only reacted to `Write` events. In the two most common scenarios this means the reload never actually happens:

1. **Kubernetes ConfigMap volumes** are updated with a symlink swap (`..data` atomically switched to a new directory), which replaces the inode — no `Write` event ever fires on the new file;
2. **Editors** saving with write-tmp-then-rename have the same effect.

So with ConfigMap-based rules (the standard deployment), `watch_rules: true` never reloads and users must restart the pods for rule changes to take effect.

This PR makes the watcher robust:

- also watch the **parent directory** of each rules file (directory events reveal the replacement),
- react to `Write`/`Create`/`Remove`/`Rename` events (filtered to the rules files or their directory),
- **re-register the file watch** after each event, since the inode may have been replaced.

The existing debounce (1s) and parse/validate/swap flow is unchanged.

**How I tested it:**

- image built from this branch deployed on a 3-node kubeadm cluster with rules mounted from a ConfigMap (directory mount);
- edited the ConfigMap → talon logged `changes detected` + `N rules have been successfully loaded` within ~20s, **zero pod restarts**;
- triggered a matching falco event → the freshly reloaded rule executed its action successfully.

Signed-off-by: gitlayzer <gitlayzer@users.noreply.github.com>

### CODE REVIEW

File: cmd/server.go [hunk 1]
```diff
@@ -84,39 +85,70 @@ var serverCmd = &cobra.Command{
...
+							utils.PrintLog(utils.ErrorStr, utils.LogLine{Error: err.Error(), Message: rulesStr})
+							return
+						}
+						// the parent directory is watched too (best effort):
+						// Kubernetes updates ConfigMap volumes with a symlink
+						// swap and editors save with write-tmp-then-rename,
+						// both replace the inode the file watch was on, and
+						// only directory events reveal the replacement.
+						if err := watcher.Add(filepath.Dir(i)); err != nil {
+							utils.PrintLog(utils.InfoStr, utils.LogLine{Error: err.Error(), Message: rulesStr})
```
Comment (Issif) [inline cmd/server.go]: **suggestion:** 


```suggestion
							utils.PrintLog(utils.ErrorStr, utils.LogLine{Error: err.Error(), Message: rulesStr})
```

**Review by Issif — CHANGES_REQUESTED** [review]

**Comment by gitlayzer** [comment]
One important deployment note for this watcher fix to be effective end-to-end:

**The rules must be mounted as a directory (not via `subPath`).**

- The official `falco-talon` chart in `falcosecurity/charts` currently mounts `rules.yaml` with `subPath` — and kubelet **never updates subPath-mounted files** when the ConfigMap changes, so no filesystem event is ever produced and the watcher (even with this fix) has nothing to react to.
- With a directory mount (e.g. mount the ConfigMap to `/etc/talon/rules` and run with `-r /etc/talon/rules/rules.yaml`), the ConfigMap volume update propagates and this PR's watcher catches it: rules reload in ~20s with zero pod restarts (verified on a 3-node kubeadm cluster).

So for upstream users to actually benefit from hot reload, the chart would also need a small companion change (mount the rules ConfigMap as a directory and point `-r` at the file inside it). I'm happy to open that companion PR on `falcosecurity/charts` if you think it's the right direction.

File: cmd/server.go [hunk 2]
```diff
@@ -84,39 +85,70 @@ var serverCmd = &cobra.Command{
...
-				for _, i := range config.RulesFiles {
-					if err := watcher.Add(i); err != nil {
-						utils.PrintLog(utils.ErrorStr, utils.LogLine{Error: err.Error(), Message: rulesStr})
-						return
+				watchRulesFiles := func() {
+					for _, i := range config.RulesFiles {
+						if err := watcher.Add(i); err != nil {
+							utils.PrintLog(utils.ErrorStr, utils.LogLine{Error: err.Error(), Message: rulesStr})
+							return
+						}
```
Comment (Issif) [inline cmd/server.go]: **Point 1 — the early `return` skips the remaining rules files.**

`return` exits the whole closure, not just this iteration. Since `watchRulesFiles()` now runs after every event, a transient `ENOENT` on one file (very likely mid-swap) leaves the other entries of `config.RulesFiles` un-refreshed, silently.

The priority should also flip: since this PR, the **directory** watch is the reliable mechanism — it's the one that sees the swap. The file watch is now just an opportunistic shortcut for the plain-write case, so failing to set it is a warning, not an error.

```suggestion
						// best effort: the inode may be gone while a swap is in
						// flight, the directory watch below is the reliable one
						if err := watcher.Add(i); err != nil {
							utils.PrintLog(utils.WarningStr, utils.LogLine{Error: err.Error(), Message: rulesStr})
						}
```

File: cmd/server.go [hunk 3]
```diff
@@ -84,39 +85,70 @@ var serverCmd = &cobra.Command{
...
+							continue
+						}
+						if !event.Has(fsnotify.Write) && !event.Has(fsnotify.Create) && !event.Has(fsnotify.Remove) && !event.Has(fsnotify.Rename) {
+							continue
+						}
+						ignore = true
+						go func() {
+							time.Sleep(1 * time.Second)
+							ignore = false
+						}()
```
Comment (Issif) [inline cmd/server.go]: **Point 2 — reload on the trailing edge instead.**

Not a one-click suggestion because it also needs `ignore := false` on line 81 to go, and that line is outside this diff hunk — applying half of it wouldn't compile. Here's the shape:

```go
// a ConfigMap swap emits a burst of events, and "..data" still points
// to the old directory when the first one arrives. The reload is
// therefore debounced on the trailing edge: the timer is rearmed on
// every event and only fires once the burst is over.
// a nil channel blocks forever in select, so nothing is pending here.
var debounce <-chan time.Time

for {
	select {
	case event := <-watcher.Events:
		// ... same filtering as today ...
		if !watched || !isRelevantOp(event) {
			continue
		}
		debounce = time.After(1 * time.Second) // rearm, don't reload

	case <-debounce:
		debounce = nil
		watchRulesFiles()
		reload() // the current body, extracted into a closure

	case err := <-watcher.Errors:
		utils.PrintLog(utils.ErrorStr, utils.LogLine{Error: err.Error(), Message: rulesStr})
	}
}
```

Three things fall out of this for free:

1. we read the filesystem **once**, after the swap has settled, instead of once before it and never again;
2. a 6-event burst causes exactly one reload instead of one reload plus five dropped events;
3. **`ignore` disappears, and with it a data race** — it's a plain `bool` written by the `Sleep` goroutine and read by this loop. `go test -race` would flag it today. Pre-existing, but this PR turns it into the primary guard, so it's worth killing here.

While we're at it: `rules` is a package-level var swapped here and read by the HTTP handlers, which is a second, larger race. Out of scope for this PR, but worth a separate issue.

File: cmd/server.go [hunk 4]
```diff
@@ -84,39 +85,70 @@ var serverCmd = &cobra.Command{
...
+								utils.PrintLog(utils.ErrorStr, utils.LogLine{Error: invalidRulesStr, Message: rulesStr})
+								break
+							}
+							utils.PrintLog(utils.InfoStr, utils.LogLine{Result: fmt.Sprintf("%v rules have been successfully loaded", len(*newRules)), Message: rulesStr})
+							rules = newRules
+							if err := actionners.Init(); err != nil {
+								utils.PrintLog(utils.ErrorStr, utils.LogLine{Error: err.Error(), Message: actionnersStr})
+								break
 							}
 						}
```
Comment (Issif) [inline cmd/server.go]: **Cleanup — the `newRules != nil` branch is now dead.**

Once the `newRules == nil` early exit was pulled out above, this second check is always true. Also worth noting that these `break`s exit the `select`, not the `for` — they behave like `continue`, which *is* what we want, but it's a well-known Go footgun and being explicit costs nothing.

```suggestion
						if newRules == nil {
							utils.PrintLog(utils.ErrorStr, utils.LogLine{Error: invalidRulesStr, Message: rulesStr})
							continue
						}
						if !validateRules(newRules) {
							utils.PrintLog(utils.ErrorStr, utils.LogLine{Error: invalidRulesStr, Message: rulesStr})
							continue
						}
						utils.PrintLog(utils.InfoStr, utils.LogLine{Result: fmt.Sprintf("%v rules have been successfully loaded", len(*newRules)), Message: rulesStr})
						rules = newRules
						if err := actionners.Init(); err != nil {
							utils.PrintLog(utils.ErrorStr, utils.LogLine{Error: err.Error(), Message: actionnersStr})
							continue
						}
```

File: cmd/server.go [hunk 5]
```diff
@@ -84,39 +85,70 @@ var serverCmd = &cobra.Command{
...
-								utils.PrintLog(utils.ErrorStr, utils.LogLine{Error: invalidRulesStr, Message: rulesStr})
+						if ignore {
+							continue
+						}
+						// only events about the rules files or their directory
+						// (symlink swaps like the ConfigMap "..data" one happen
+						// in the same directory)
+						watched := false
+						for _, i := range config.RulesFiles {
+							if event.Name == i || filepath.Dir(event.Name) == filepath.Dir(i) {
```
Comment (Issif) [inline cmd/server.go]: **nit, your call** — this makes *any* event in the rules directory trigger a reload. Intended for `..data`, but if someone runs `-r /etc/talon/rules.yaml` with `talon.yaml` sitting in the same directory, every config write reloads the rules. (The `event.Name == i` branch is also already subsumed by the second one as written.)

Narrowing to the swap marker still covers all three scenarios: direct write and editor rename both surface as `event.Name == i`, ConfigMap swaps as `..data`.

```suggestion
							if event.Name == i || (filepath.Base(event.Name) == "..data" && filepath.Dir(event.Name) == filepath.Dir(i)) {
```

**Review by Issif — CHANGES_REQUESTED** [review]
Thanks for digging into this — the diagnosis is spot on and it's a real bug: `inotify` follows symlinks, so `watcher.Add()` on the rules file actually watches the inode of `..2026_..._xx/rules.yaml`, which kubelet deletes on every ConfigMap update. Watching the parent directory plus re-registering the file watch is the right fix.

Two things I'd like sorted before this goes in, both correctness rather than style, plus a couple of cleanups.

**The debounce fires on the wrong edge (main one).** The reload runs on the *first* event of the burst, and a ConfigMap swap emits its events roughly like this:

| t | event | `..data` points to |
|---|---|---|
| 0 ms | `Create ..2026_09_04_13_05_00.222` | **old** dir |
| ~2 ms | `Create ..data_tmp` | old dir |
| ~3 ms | `Rename ..data_tmp` / `Create ..data` | **new** dir |
| ~5 ms | `Remove` old dir, `Remove rules.yaml` | new dir |

We reload at t=0 ms, when `..data` still resolves to the old directory, so `ParseRules` reads the **old** content — then `ignore` swallows the t=3 ms event that was the only one that mattered. Talon logs `changes detected` and `N rules have been successfully loaded` while nothing actually changed.

What makes it slippery is that it depends on a race: the `Sleep(1s)` runs concurrently with the synchronous reload, while the remaining events queue up in the buffered `watcher.Events` channel. If parse + validate + `actionners.Init()` take **more** than 1s, `ignore` is back to `false` by the time we drain the channel and a second, correct reload happens — which is likely why your cluster test passed. Under 1s, it silently keeps the stale rules. A trailing-edge debounce removes the race entirely (and makes the `ignore` data race go away with it — see the inline comment).

**`watchRulesFiles()` bails out on the first error.** This is now called after *every* event, i.e. precisely when `watcher.Add` is most likely to hit `ENOENT` on a file being swapped — and the `return` then skips the remaining entries of `config.RulesFiles`. With two ConfigMap-backed rules files, an update to the first one permanently stops refreshing the file watch of the second. Inline suggestion below.

(To be fair, the PR *improves* the blast radius here: that `return` used to kill the whole goroutine while the startup log still claimed `watch of rules enabled`. It just needs to go one step further.)

**Two more things, not blocking the review:**

- **Tests.** The watcher is an inline closure in `serverCmd.RunE`, which is exactly why this bug survived so long — nothing can exercise it. Extracting it into a `watchRules(...)` function in its own file would let us cover the symlink-swap scenario with a temp dir. I'd really like that here, since the regression is trivial to reintroduce.
- **The chart.** You're right in your comment, and it matters: the chart mounts `rules.yaml` with `subPath`, and kubelet never propagates ConfigMap updates to `subPath` mounts, so no filesystem event is produced at all. Please do open the companion PR on `falcosecurity/charts` and link it here — without it this fix is invisible to upstream users. A note in the docs about the directory mount would help too.

**Comment by gitlayzer** [comment]
Thanks @Issif for the sharp review — all four points are addressed in `432d497`:

1. **Point 1**: the early `return` is gone, one failing file no longer leaves the others unwatched; file-watch `Add` failures are now `WarningStr` (best effort), the directory watch is documented as the reliable mechanism.
2. **Point 2**: reworked to trailing-edge debounce exactly in the shape you sketched (`var debounce <-chan time.Time`, rearm on every event, fire once after the burst settles). The `ignore` bool — and its race — is gone.
3. **Cleanup**: the dead `newRules != nil` second check is removed, the reload body is extracted into a closure, and `break`→`continue`/`return` everywhere.
4. **Nit**: filter narrowed to the rules file itself or the `..data` swap marker — other files in the same directory no longer trigger reloads.

Also filed #800 for the larger `rules` variable race you flagged (handlers reading while the watcher swaps) — agreed it's out of scope here but shouldn't be lost.

**Comment by gitlayzer** [comment]
All four points addressed and CI green on the branch:

- **Trailing-edge debounce** (the main correctness issue) — implemented exactly in the shape you sketched, `ignore` and its race are gone
- **`watchRulesFiles()` early return** — removed; file-watch failures are warnings, the directory watch is the reliable mechanism
- **Dead code / explicit continues** — done
- **Filter narrowed** to the rules file and the `..data` swap marker

Two more things you asked for, now also done:

- **Tests**: the watcher is extracted into `watchRules()` in `cmd/watch_rules.go` with the reload body injected, and covered in `cmd/watch_rules_test.go` — including `TestWatchRulesConfigMapSwap`, a temp-dir kubelet-style symlink swap that asserts the reload fires **and observes the new content** (passes on Linux; scenario tests are Linux-gated since kqueue doesn't emit directory events).
- **Chart companion PR**: opened → falcosecurity/charts#1059 (mounts the rules ConfigMap as a directory at `/etc/falco-talon/rules.d`, chart 0.5.0) — without it this fix is indeed invisible to upstream users.

@Issif PTAL when you have a moment 🙏

File: cmd/watch_rules_test.go [hunk 6]
```diff
@@ -0,0 +1,149 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package cmd
+
+// 场景测试（ConfigMap 交换/重命名保存）依赖 inotify 的目录事件语义，
+// 仅在 Linux 上运行（macOS 的 kqueue 后端不产生目录内容事件）。
```
Comment (Issif) [inline cmd/watch_rules_test.go]: This needs to be in English — the rest of the tree is, and it's the convention upstream.

```suggestion
// The scenario tests (ConfigMap swap / rename save) rely on inotify's
// directory event semantics and only run on Linux (the kqueue backend on
// macOS does not emit directory content events).
```

File: cmd/watch_rules_test.go [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+func skipIfNotLinux(t *testing.T) {
+	t.Helper()
+	if runtime.GOOS != "linux" {
+		t.Skip("fsnotify directory events require inotify (Linux)")
+	}
+}
+
+// setupConfigMapLayout builds a temp dir mimicking a Kubernetes ConfigMap
+// volume: rules.yaml is a symlink chain into a versioned "..data" dir.
+func setupConfigMapLayout(t *testing.T, version string) (root, rulesFile string) {
```
Comment (Issif) [inline cmd/watch_rules_test.go]: `version` is never read — the body hardcodes `v1`, and the only call site passes `"v1"`. None of the enabled linters catch it (`unparam` isn't on, and revive's `unused-parameter` isn't in the default rule set), so it'd just sit there.

```suggestion
func setupConfigMapLayout(t *testing.T) (root, rulesFile string) {
```

The call on line 79 needs the argument dropped to match.

File: cmd/watch_rules_test.go [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+	t.Helper()
+	root = t.TempDir()
+	v1 := filepath.Join(root, "v1")
+	require.NoError(t, os.MkdirAll(v1, 0o755))
+	require.NoError(t, os.WriteFile(filepath.Join(v1, "rules.yaml"), []byte("v1"), 0o644))
```
Comment (Issif) [inline cmd/watch_rules_test.go]: `gosec` G301 + G306, failing the lint job (and with it the whole test job).

```suggestion
	require.NoError(t, os.MkdirAll(v1, 0o750))
	require.NoError(t, os.WriteFile(filepath.Join(v1, "rules.yaml"), []byte("v1"), 0o600))
```

File: cmd/watch_rules_test.go [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+	return root, rulesFile
+}
+
+// swapConfigMap simulates a kubelet ConfigMap update: build the new version
+// dir, then flip the "..data" symlink via a tmp rename, like kubelet does.
+func swapConfigMap(t *testing.T, root, content string) {
+	t.Helper()
+	v2 := filepath.Join(root, "v2")
+	require.NoError(t, os.MkdirAll(v2, 0o755))
+	require.NoError(t, os.WriteFile(filepath.Join(v2, "rules.yaml"), []byte(content), 0o644))
```
Comment (Issif) [inline cmd/watch_rules_test.go]: Same G301 + G306 here.

```suggestion
	require.NoError(t, os.MkdirAll(v2, 0o750))
	require.NoError(t, os.WriteFile(filepath.Join(v2, "rules.yaml"), []byte(content), 0o600))
```

The other four G306s are on lines 110, 118, 132 and 141 — same `0o644` → `0o600` change.

File: cmd/watch_rules_test.go [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+	skipIfNotLinux(t)
+	root, rulesFile := setupConfigMapLayout(t, "v1")
+
+	var calls atomic.Int32
+	var seen atomic.Value
+	done := make(chan struct{})
+	var once sync.Once
+	reload := func() {
+		calls.Add(1)
+		b, _ := os.ReadFile(rulesFile)
```
Comment (Issif) [inline cmd/watch_rules_test.go]: G304. The path comes from `t.TempDir()` so it's fine in substance — just annotate it rather than restructure.

```suggestion
		b, _ := os.ReadFile(rulesFile) //nolint:gosec // test-controlled temp path
```

File: cmd/watch_rules_test.go [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+	// give the watcher a moment to register
+	time.Sleep(300 * time.Millisecond)
+	swapConfigMap(t, root, "v2")
+
+	select {
+	case <-done:
+	case <-time.After(5 * time.Second):
+		t.Fatal("no reload happened within 5s after the ConfigMap swap")
+	}
+	require.Equal(t, "v2", seen.Load(), "reload must observe the swapped-in content")
```
Comment (Issif) [inline cmd/watch_rules_test.go:103]: **This test passes with a leading-edge debounce too, so it doesn't protect the fix it was written for.**

I checked by putting the old behaviour back locally:

```go
if debounce == nil {
	debounce = time.After(1 * time.Second)
	watchRulesFiles()
	reload()   // reload on the FIRST event, the old bug
}
```

`--- PASS: TestWatchRulesConfigMapSwap (0.31s)`.

Why: `isRulesEvent` filters out the `v2` directory creation and the `..data_tmp` creation, so the first event reaching the debounce is already the `..data` flip — the new content is visible by then, and `seen` is `"v2"` either way. The narrowed filter (a good change on its own) happens to mask the distinction the test is trying to measure.

What separates the two edges is the *number* of reloads over a burst, which is what `calls` on line 86 is already tracking but nothing asserts. Something like:

```go
// the swap emits a burst; the trailing edge must collapse it into one reload
time.Sleep(2 * time.Second) // debounce is 1s, let it settle
require.Equal(t, int32(1), calls.Load(), "the burst must produce exactly one reload")
```

A leading-edge implementation reloads at t=0 *and* leaves the rearmed timer to fire, so it lands on 2 and the test fails. Worth also softening the claim on line 76 to match what's actually asserted.

File: cmd/watch_rules.go [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+
+// swapMarker is the symlink name Kubernetes flips on ConfigMap volume updates.
+const swapMarker = "..data"
+
+// watchRules watches the given rules files (and their parent directories)
+// and invokes reload when they change. reload is debounced on the trailing
+// edge: a ConfigMap swap emits a burst of events while "..data" still points
+// to the old directory when the first one arrives, so the timer is rearmed
+// on every event and only fires once the burst is over.
+func watchRules(rulesFiles []string, reload func()) {
```
Comment (Issif) [inline cmd/watch_rules.go]: nit, non-blocking — no way to stop this. The `for` loop never returns and there's no `ctx`/`done`, so each of the three scenario tests leaks a goroutine that goes on watching a `t.TempDir()` that the cleanup has already removed. Harmless today, but since the whole point of the extraction was testability, a cancellation parameter would round it off (and let the tests assert the watcher shuts down rather than abandoning it).

File: cmd/watch_rules.go [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		utils.PrintLog(utils.ErrorStr, utils.LogLine{Error: err.Error(), Message: rulesStr})
+		return
+	}
+	defer func() { _ = watcher.Close() }()
```
Comment (Issif) [inline cmd/watch_rules.go:29]: **Unclean rules paths silently stop reloading — this is a regression against `main`.**

`isRulesEvent` compares `event.Name == i` as exact strings, but fsnotify reports event names in cleaned form. Reproduced with `-r <dir>/./rules.yaml`:

```
--- FAIL: TestWatchRulesUncleanPathDirectWrite (3.30s)
    no reload after a direct write with an unclean rules path
```

No reload at all. On `main` this worked, because there was no name filtering — any `Write` triggered a reload. The ConfigMap path is unaffected (the `..data` branch compares `filepath.Dir`, which cleans), and `mage run` passes `-r rules.yaml`, which is already clean — but `-r ./rules.yaml` is an easy thing for someone to type locally.

Cleaning once up front fixes it; I ran the suite with this applied and all four tests pass:

```suggestion
	defer func() { _ = watcher.Close() }()

	// fsnotify reports event names in cleaned form, so the paths the
	// filter compares against have to be cleaned too.
	files := make([]string, 0, len(rulesFiles))
	for _, i := range rulesFiles {
		files = append(files, filepath.Clean(i))
	}
	rulesFiles = files
```

File: cmd/server.go [hunk 14]
```diff
@@ -76,55 +74,25 @@ var serverCmd = &cobra.Command{
...
-									utils.PrintLog(utils.ErrorStr, utils.LogLine{Error: err.Error(), Message: actionnersStr})
-									break
-								}
-							}
-						}
-					case err := <-watcher.Errors:
-						utils.PrintLog(utils.ErrorStr, utils.LogLine{Error: err.Error(), Message: rulesStr})
-					}
+				utils.PrintLog(utils.InfoStr, utils.LogLine{Result: fmt.Sprintf("%v rules have been successfully loaded", len(*newRules)), Message: rulesStr})
+				rules = newRules
```
Comment (Issif) [inline cmd/server.go]: nit, non-blocking — this write is dead. `rules` is the local declared on line 39 in `Run`; nothing reads it after startup (the `validateRules` and the log both run earlier), and `ruleengine.ParseRules` has already published the package-level var by the time we get here. It came across from the old code, so not something this PR introduced.

Dropping it would make it clearer that the swap happens inside `ParseRules` — which is also the point #800 is about.

**Review by Issif — COMMENTED** [review]
Re-reviewed at `61e8538`. All six points from the last pass are addressed, and the trailing-edge rework is exactly the right shape — thanks for turning that around so fast. `go test -race ./cmd/...` is clean on my side and the three scenario tests do genuinely exercise the swap.

Three things still block, one of which is more interesting than it looks.

### 1. CI is red — lint fails, and the tests never ran

I kicked off the workflow (it was sitting in `action_required`, first-contribution gate). The run on this SHA came back `failure`:

```
cmd/watch_rules_test.go:34:21: G301: Expect directory permissions to be 0750 or less (gosec)
cmd/watch_rules_test.go:35:21: G306: Expect WriteFile permissions to be 0600 or less (gosec)
cmd/watch_rules_test.go:47:21: G301: Expect directory permissions to be 0750 or less (gosec)
cmd/watch_rules_test.go:48:21: G306: Expect WriteFile permissions to be 0600 or less (gosec)
cmd/watch_rules_test.go:87:11: G304: Potential file inclusion via variable (gosec)
cmd/watch_rules_test.go:110:21: G306: Expect WriteFile permissions to be 0600 or less (gosec)
cmd/watch_rules_test.go:118:21: G306: Expect WriteFile permissions to be 0600 or less (gosec)
cmd/watch_rules_test.go:132:21: G306: Expect WriteFile permissions to be 0600 or less (gosec)
cmd/watch_rules_test.go:141:21: G306: Expect WriteFile permissions to be 0600 or less (gosec)
9 issues
```

`.golangci.yml` excludes `_test.go` from `bodyclose` and `goconst`, but not from `gosec`. And since `pr.yaml` declares `test: needs: lint`, the `test` job was **skipped** — so the new tests have not executed in CI even once. Worth fixing the lint first and letting the suite actually run, rather than trusting local results.

Inline suggestions below for the permissions and the `G304`.

### 2. `TestWatchRulesConfigMapSwap` does not guard the regression it was written for

This is the one I'd really like sorted. I reintroduced a leading-edge debounce locally — reload on the first event, subsequent ones swallowed for 1s, i.e. precisely the bug from the last review — and **the test still passes**, in 0.31s instead of 1.30s.

The cause is the narrowed filter from point 4 of the last pass. `isRulesEvent` now drops every pre-swap event — the `v2` directory creation and the `..data_tmp` creation — so the first event that survives the filter is the `..data` flip itself, at which point the new content is already visible. Leading and trailing edge become indistinguishable on this scenario, and the content assertion cannot tell them apart.

That doesn't make the trailing-edge design wrong: it still coalesces bursts, and it's what removed the `ignore` race. But the comment on line 76 claims more than the test demonstrates, and a refactor back to the leading edge would sail through green. Note that `calls` is incremented on line 86 and never asserted — counting reloads across a burst is what would actually pin the behaviour. See the inline comment.

### 3. The comment on lines 5-6 is in Chinese

Needs to be English, like the rest of the tree.

---

**Also found while testing: a regression against `main` on unclean paths.** Detail and a one-line fix inline on `cmd/watch_rules.go`.

Two loose ends from my side, not asks on you:

- **The docs note I asked for last time — drop it.** That was a bad ask: there's nothing in this repo but `config_example.yaml:7`, the real docs live in a separate repo. I'll handle it there.
- **The chart PR** I'll review on `falcosecurity/charts` directly, and it won't merge before the next falco-talon release ships anyway. One heads-up so it isn't a surprise: `appVersion` is still `0.3.0` there, and `image.tag` defaults to `.Chart.AppVersion`, so as it stands chart 0.5.0 would render `falco-talon:0.3.0` — the release *without* this watcher fix — while the CHANGELOG promises hot reload works. That bump needs to land in the same PR.

File: cmd/server.go (hunk 1 above)
Comment (Issif) [reply cmd/server.go]: Addressed in `61e8538` — the directory watch failure now logs with `ErrorStr` at `cmd/watch_rules.go:42`. Resolving.

File: cmd/server.go (hunk 2 above)
Comment (Issif) [reply cmd/server.go]: Addressed in `61e8538` — `cmd/watch_rules.go:30-45`: the early `return` is gone so one failing file no longer leaves the others unwatched, `watcher.Add(i)` failures log at `WarningStr` (best effort), and the directory watch keeps `ErrorStr` as the reliable mechanism. Resolving.

File: cmd/server.go (hunk 3 above)
Comment (Issif) [reply cmd/server.go]: Addressed in `61e8538` — `cmd/watch_rules.go:50-64` implements the trailing-edge debounce in the shape sketched here, and `ignore` (with its data race) is gone. Resolving this thread; the follow-up that the test doesn't actually pin the edge is tracked in its own thread on `cmd/watch_rules_test.go:102`, and the `rules` package-var race in #800.

File: cmd/server.go (hunk 4 above)
Comment (Issif) [reply cmd/server.go]: Addressed in `61e8538` — `cmd/server.go:79-92`: the dead `newRules != nil` branch is gone, and since the reload body is now an injected closure the exits are explicit `return`s rather than `break`s. Resolving.

File: cmd/server.go (hunk 5 above)
Comment (Issif) [reply cmd/server.go]: Addressed in `61e8538` — `isRulesEvent` at `cmd/watch_rules.go:74-84` narrows the filter to the rules file itself or the `..data` marker in its directory, so unrelated writes alongside it no longer trigger a reload. Resolving.

**Comment by gitlayzer** [comment]
Third round addressed in `fbd83a4` — thank you for the precision, especially #2.

**Blocking:**

1. **CI red (gosec)**: all 9 issues fixed — `0o750`/`0o600` perms throughout the test file and `//nolint:gosec` on the test-controlled `ReadFile`. My mistake for trusting local results while `test: needs: lint` meant the suite never actually ran in CI.
2. **Edge pinning**: the test now asserts **exactly one reload across the burst** (`calls == 1` after a 2s settle), not just the observed content. Verified that a leading-edge implementation lands on 2 and fails this — the regression you reintroduced locally is now actually guarded.
3. **English comment** — done, verbatim with your suggestion.

**Regression vs `main`**: fixed with your one-liner — `filepath.Clean` on the rules paths up front, plus `TestWatchRulesUncleanPathDirectWrite` covering `-r ./rules.yaml`. Confirmed main worked here only because it had no name filter at all; good catch that the narrowed filter reintroduced it.

**Nits**: unused `version` param dropped, `watchRules` takes a `done` channel (with `TestWatchRulesCancellation` asserting no reload fires after close), and the dead `rules = newRules` write is gone — you're right, `ParseRules` publishes the package var at `internal/rules/rules.go:231`, which also reinforces #800's framing.

All 6 watcher tests pass on Linux (`TestIsRulesEvent`, ConfigMapSwap, DirectWrite, UncleanPathDirectWrite, RenameSave, Cancellation), `go test -race ./cmd/...` clean, gosec reports 0 issues.

**Chart PR**: `appVersion` bumped to `0.3.1` in `53bf448` per your heads-up — good spot, chart 0.5.0 would indeed have rendered the pre-fix image otherwise.

@Issif PTAL 🙏

**Comment by gitlayzer** [comment]
Closing this PR — we've decided to keep this feature for internal use rather than pursuing the upstream contribution at this time. Thank you @Issif for the exceptionally thorough reviews across three rounds — they made the code genuinely better (the trailing-edge debounce design, the edge-pinning test, the unclean-path catch). The branch stays at gitlayzer/falco-talon (feat/rules-watch-configmap) if anyone wants to pick it up. #800 (rules var race) remains valid independently.

**Comment by Issif** [comment]
I was interesting for me too, thanks. I'm glad to know your company is using Talon :pray:
